File: exercise/leetcode/0042/0042.cc
#include <vector>
#include <algorithm>
#include <stack>
using namespace std;

class Solution {
public:
    int trap(vector<int>& height) {
        stack<int> st;
        int cnt = 0;
        for (int i = 0; i < height.size(); ++i) {
            while (!st.empty() && height[st.top()] < height[i]) {
                int a = height[st.top()];
                st.pop();
                if (st.empty()) continue;
                int li = st.top();
                int mh = min(height[i], height[li]);
                cnt += (mh - a) * (i - li - 1);
            }
            st.push(i);
        }
        return cnt;
    }
};

File: exercise/leetcode/0042/0042.py
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        st = []
        ln = len(height)
        cnt = 0
        for i in range(ln):
            while len(st) > 0 and height[st[-1]] < height[i]:
                d = height[st[-1]]
                st.pop()
                if len(st) == 0:
                    continue
                li = st[-1]
                mh = min(height[li], height[i])
                cnt += (mh - d) * (i - li - 1)
            st.append(i)

        return cnt